### PR TITLE
[DPE-2393] Common anchestor to unify shared characteristics of DataRequires and DataProvides

### PR DESCRIPTION
**Pull request by juditnovak**

### CODE REVIEW

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 1]
```diff
@@ -430,13 +418,45 @@ def _update_relation_data(self, relation_id: int, data: dict) -> None:
                 that should be updated in the relation.
         """
         if self.local_unit.is_leader():
-            if relation := self.charm.model.get_relation(self.relation_name, relation_id):
+            relation = self.charm.model.get_relation(self.relation_name, relation_id)
+            if relation:
```
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:422]: `lint` made me do this.I mean to add the check (against `None`)

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 2]
```diff
@@ -430,13 +418,45 @@ def _update_relation_data(self, relation_id: int, data: dict) -> None:
...
+
     @property
     def relations(self) -> List[Relation]:
         """The list of Relation instances associated with this relation_name."""
-        return list(self.charm.model.relations[self.relation_name])
+        return [
+            relation
+            for relation in self.charm.model.relations[self.relation_name]
+            if self._is_relation_active(relation)
+        ]
```
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:440]: I'm planning to add some secret-related functions to the new, common anchestor.
Typically like checking for Juju version

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 3]
```diff
@@ -938,11 +895,8 @@ def _assign_relation_alias(self, relation_id: int) -> None:
 
         # Return if an alias was already assigned to this relation
         # (like when there are more than one unit joining the relation).
-        if (
-            self.charm.model.get_relation(self.relation_name, relation_id)
-            .data[self.local_unit]
-            .get("alias")
-        ):
+        relation = self.charm.model.get_relation(self.relation_name, relation_id)
+        if relation and relation.data[self.local_unit].get("alias"):
```
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:899]: I double-checked both in Kafka and Opensearch. This function is never called outside of this particular module here.
Thus consistently changing this function's name (for consistency and common inheritance) I supposed to be a transparent operation.

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 4]
```diff
@@ -955,7 +909,8 @@ def _assign_relation_alias(self, relation_id: int) -> None:
 
         # Set the alias in the unit relation databag of the specific relation.
         relation = self.charm.model.get_relation(self.relation_name, relation_id)
-        relation.data[self.local_unit].update({"alias": available_aliases[0]})
+        if relation:
+            relation.data[self.local_unit].update({"alias": available_aliases[0]})
```
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:913]: `lint` made me add this check.

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 5]
```diff
@@ -797,7 +754,7 @@ class DatabaseProvides(DataProvides):
     def __init__(self, charm: CharmBase, relation_name: str) -> None:
         super().__init__(charm, relation_name)
 
-    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+    def _on_relation_changed_event(self, event: RelationChangedEvent) -> None:
```
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:757]: I double-checked both in Kafka and Opensearch. This function is never called outside of this particular module here.
Thus consistently changing this function's name (for consistency and common inheritance) I supposed to be a transparent operation.

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 3 above)
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:899]: `lint` made me add this check against `None`

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 6]
```diff
@@ -379,35 +379,23 @@ def __init__(self, charm: CharmBase, relation_name: str) -> None:
         self.relation_name = relation_name
         self.framework.observe(
             charm.on[relation_name].relation_changed,
-            self._on_relation_changed,
+            self._on_relation_changed_event,
```
Comment (juditnovak) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:382]: I double-checked both in Kafka and Opensearch. This function is never called outside of this particular module here.
Thus consistently changing this function's name (for consistency and common inheritance) I supposed to be a transparent operation.

(Pls let me know if any other charm may be known to potentially directly use this method by any chance)

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 7]
```diff
@@ -368,8 +368,8 @@ def diff(event: RelationChangedEvent, bucket: Union[Unit, Application]) -> Diff:
 # Base DataProvides and DataRequires
```
Comment (welpaolo) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:368]: Move this line after the new class

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 8]
```diff
@@ -430,13 +419,45 @@ def _update_relation_data(self, relation_id: int, data: dict) -> None:
                 that should be updated in the relation.
         """
         if self.local_unit.is_leader():
-            if relation := self.charm.model.get_relation(self.relation_name, relation_id):
+            relation = self.charm.model.get_relation(self.relation_name, relation_id)
+            if relation:
                 relation.data[self.local_app].update(data)
 
+    @staticmethod
+    def _is_relation_active(relation: Relation):
```
Comment (welpaolo) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:427]: Can you add docstring please ?

**Review by marceloneppel — APPROVED** [review]
LGTM! Just don't forget to bum the LIBPATCH field as Paulo mentioned.

**Pull request merged**